homework/console/02/control.ts

```diff
@@ -0,0 +1,108 @@
+interface DoingInterface {
+  category: string;
+  content: string[];
+}
+interface MemoryInterface {
+  name: string;
+  gender: string;
+  age: number;
+  isStudent: boolean;
+  doing?: DoingInterface[];
+  hobby?: string[];
+}
+type ConvertedGender = { [key: string]: string };
+type actionType = MemoryInterface | string | number;
+
+const convertGender: ConvertedGender = {
+  female: '여성',
+  male: '남성',
+};
+
+const arr: number[] = [];
+const control = (type: string, action: actionType) => {
+  let gameStatus = '';
+  let result: string | number[] = '';
+  if (type === 'game') {
+    switch (action) {
+      case 'start':
+        gameStatus = '게임이 시작되었습니다!';
+        break;
+      case 'pause':
+        gameStatus = '게임이 중지 되었습니다!';
+        break;
+      case 'stop':
+        gameStatus = '게임이 종료 되었습니다!';
+        break;
+    }
+    result = gameStatus;
+  }
+
+  if (type === 'study') {
+    if (action > 0) {
+      !arr.includes(+action) && arr.push(+action);
+    } else {
+      !arr.includes(+action) && arr.pop();
+    }
+    result = arr;
+  }
+
+  if (
+    type === 'memory' &&
+    typeof action !== 'string' &&
+    typeof action !== 'number'
+  ) {
+    result = `저의 이름은 ${action.name}, ${convertGender[action.gender]}이고 ${
```

Comment:
아래와 같이 제네릭을 활용해볼수도 있을것 같아요!
```suggestion
if (type === 'memory') {
    const memory = <MemoryInterface>action;

    result = `저의 이름은 ${memory.name}, ${convertGender[memory.gender]}이고
```

Reply:
와.... 이거 진짜... 리뷰 받길 바란 부분이었는데 제가 기다리던 완벽한 리뷰 감사합니다! 🥰
감사합니다 민오 👍🏻 체고체고

Reply:
> 와.... 이거 진짜... 리뷰 받길 바란 부분이었는데 제가 기다리던 완벽한 리뷰 감사합니다! 🥰 감사합니다 민오 👍🏻 체고체고

이제는 if에서 type 으로만 확인하니 로직을 분리해서 함수로 추상화 할수 있을 것 같아요!

Reply:
그 혹시 알고 계실지도 모르지만 첨부 하자면

```ts
type actionType = MemoryInterface | string | number;
```

```ts
const control = (type: GameType, action: actionType)
```

```ts
const memory = <MemoryInterface>action;
```

actionType 에 MemoryInterface 타입이 있어서  추론이 가능하기 때문에 정상 동작 해요!

Reply:
오.. 위에서 저렇거 선언되었기 때문에 정상동작 하는거군요!! 감사합니다 저렇게 deep 하게는 생각 못했네요 😊